[AArch64] Disable FEAT_RNG on Grace.

rj-jesus opened the pull request:
The FIXME in the test changed should be removed by #176340.

davemgreen reviewed (commented):
This should be left on. People can use -mcpu=neoverse-v2+norng if they don't have it. Lots of features can be disabled for many different reasons. If Grace doesn't have it then maybe that should apply just to Grace, and we split the configs into two?

The mcpu=native change looks sensible. We should be doing that for all features AFAIU (and maybe getting them from hwcaps).

rj-jesus commented:
> This should be left on. People can use -mcpu=neoverse-v2+norng if they don't have it. Lots of features can be disabled for many different reasons.

Hi, I assume the intent of having it on is to enable the usage of the random number generator intrinsics gated by some form of runtime detection? If so, should that not still be possible with `-mcpu=neoverse-v2+rng`, which might better convey that the runtime detection is indeed _necessary_ and shouldn't be presumed?

AFAICT the availability of the rng intrinsics in CPUs that do not support them is not specified (or at least I couldn't find it), which can lead to attempts to execute illegal instructions from what looks like reasonable source code. From the [ACLE](https://github.com/ARM-software/acle/blob/main/main/acle.md#random-number-generation-extension):
> __ARM_FEATURE_RNG is defined to 1 if the Random Number Generation instructions are supported and the intrinsics defined in [Random number generation intrinsics](https://github.com/ARM-software/acle/blob/main/main/acle.md#random-number-generation-intrinsics) are available.

There doesn't seem to be any other conditions under which __ARM_FEATURE_RNG is defined, which---unless I'm misunderstanding or missing something---seems to suggest the instructions should be available unconditionally alongside the intrinsics.

> If Grace doesn't have it then maybe that should apply just to Grace, and we split the configs into two?

Yeah, I forgot to mention it earlier, the problem with going this route is that it will still misbehave with `-mcpu=native`. AFAIU `getHostCPUFeatures` is mainly used to enable features when we find them, not to disable them when we don't (at least I couldn't do so trivially when I gave it a try). If we go this route, we may need to add some means for excluding extensions like `rng`, or define a "base" Neoverse V2 spec without crypto and rng and use that with `-mcpu=native`. Neither option sounded great, but please let me know if there's something else I should try.

rj-jesus committed:
[AArch64][Driver] Allow runtime detection to override default feaures.

Currently, most extensions controlled through -march and -mcpu options
are handled in a bitset of AArch64::ExtensionSet. However, extensions
detected at runtime for native compilation are handled in a separate
list of CPU features; once most of the parsing logic has run, the bitset
is converted to a list of features, added after the features detected at
runtime, and the resulting list is used from there on.

This has the downside that runtime-detected features are unable to
override default CPU extensions. For example, if a CPU enables +aes in
its processor definition, but aes is not found at runtime, the feature
currently remains enabled---even though unsupported---because default
features are enabled after the runtime logic attempts to disable them.

This patch inserts runtime-detected features directly into the extension
set such that these options can take precedence over extensions enabled
by default. The general parsing order for mcpu=native becomes:
1. CPU defaults;
2. Runtime detection;
3. Further +featureA+nofeatureB options;
4. Other parsing decisions.

This allows feature that are found to be unsupported at runtime to be
removed from the list of features supported by targets that enable them
by default.

rj-jesus committed:
[AArch64] Disable FEAT_RNG on Grace.

This also disabled rng if not found at runtime.

rj-jesus commented:
> I understood there's now consensus to go ahead with this. Just one more question, GCC was already disabling this?

Yes, this is already disabled in GCC.

davemgreen reviewed (commented):
The part that alters -mcpu=grace sounds good. I believe that can be committed separately and isn't dependant on the rest. The part that alters Host.cpp is related to -mcpu=native and could be better included of the other patch, so that -mcpu=native works correctly.

(I think how it would best work is to use hpcaps on linux and trust all the feaures we have, but that is a larger change).

rj-jesus committed:
Revert "[AArch64][Driver] Allow runtime detection to override default feaures."

This reverts commit 66450a937f415c302793901ddcb67a2ba161265c.

rj-jesus committed:
Drop Host.cpp changes.

rj-jesus commented:
> The part that alters -mcpu=grace sounds good. I believe that can be committed separately and isn't dependant on the rest. The part that alters Host.cpp is related to -mcpu=native and could be better included of the other patch, so that -mcpu=native works correctly.

Thanks, that should be done now. Depending on the PR that's merged first, I'll update the other to remove the corresponding FIXME.

> (I think how it would best work is to use hpcaps on linux and trust all the feaures we have, but that is a larger change).

Agreed, I think that's how we should be doing it. Maybe some of the FMV code in compiler-rt could be used for it?

davemgreen reviewed (approved):
Thanks - LGTM
